geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/cluster/ClusterExecutor.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.geode.redis.internal.netty.Command;
+import org.apache.geode.redis.internal.netty.ExecutionHandlerContext;
+
+public class ClusterExecutor extends AbstractExecutor {
+
+  @Override
+  public RedisResponse executeCommand(Command command, ExecutionHandlerContext context) {
+
+    List<byte[]> args = command.getProcessedCommand();
+    String subCommand = new String(args.get(1));
+
+    switch (subCommand.toLowerCase()) {
+      case "info":
+        return getInfo(context);
+      case "nodes":
+        return getNodes(context);
+      case "slots":
+        return getSlots(context);
+      default: {
+        return RedisResponse.error(
+            String.format(ERROR_UNKNOWN_CLUSTER_SUBCOMMAND, subCommand));
+      }
+    }
+  }
+
+  private RedisResponse getSlots(ExecutionHandlerContext ctx) {
+    List<MemberBuckets> memberBuckets = getMemberBuckets(ctx);
+
+    Map<Integer, String> primaryBucketToMemberMap = new HashMap<>();
+    Map<String, Pair<String, Integer>> memberToHostPortMap = new TreeMap<>();
+    int retrievedBucketCount = 0;
+
+    for (MemberBuckets m : memberBuckets) {
+      memberToHostPortMap.put(m.getMemberId(), Pair.of(m.getHostAddress(), m.getPort()));
+      for (Integer id : m.getPrimaryBucketIds()) {
+        primaryBucketToMemberMap.put(id, m.getMemberId());
+        retrievedBucketCount++;
+      }
+    }
+
+    if (retrievedBucketCount != REDIS_REGION_BUCKETS) {
+      return RedisResponse.error("Internal error: bucket count mismatch " + retrievedBucketCount
+          + " != " + REDIS_REGION_BUCKETS);
+    }
+
+    int index = 0;
+    List<Object> slots = new ArrayList<>();
+
+    for (int i = 0; i < REDIS_REGION_BUCKETS; i++) {
+      Pair<String, Integer> primaryHostAndPort =
+          memberToHostPortMap.get(primaryBucketToMemberMap.get(i));
+
+      List<Object> entry = new ArrayList<>();
+      entry.add(index * REDIS_SLOTS_PER_BUCKET);
+      entry.add(((index + 1) * REDIS_SLOTS_PER_BUCKET) - 1);
+      entry.add(Arrays.asList(primaryHostAndPort.getLeft(), primaryHostAndPort.getRight()));
+
+      slots.add(entry);
+      index++;
+    }
+
+    return RedisResponse.array(slots);
+  }
+
+  @SuppressWarnings("unchecked")
+  private List<MemberBuckets> getMemberBuckets(
+      ExecutionHandlerContext ctx) {
+    Region<RedisKey, RedisData> dataRegion = ctx.getRegionProvider().getDataRegion();
+
+    // Really only need this in situations where the cluster is empty and no data has been
+    // added yet.
+    PartitionRegionHelper.assignBucketsToPartitions(dataRegion);
+
+    Set<DistributedMember> membersWithDataRegion = new HashSet<>();
+    for (PartitionMemberInfo memberInfo : getRegionMembers(ctx)) {
+      membersWithDataRegion.add(memberInfo.getDistributedMember());
+    }
+
+    ResultCollector<MemberBuckets, List<MemberBuckets>> resultCollector =
+        FunctionService.onMembers(membersWithDataRegion).execute(BucketInfoRetrievalFunction.ID);
```

Comment:
It's somewhat unfortunate we need to use a function at all to get this information. The location of the *buckets* is actually known to all members of the PR, so we can find out where all the primaries are without a function. I think the part we don't know is the redis bind address on the other members.

Is it worth improving the performance of these commands, or are they so infrequent we don't care? If it is worth it, we can add another story to figure out how to distribute the redis bind addresses (or maybe just cache them?).

Reply:
ATM I think it's OK since these are called once when a connection is established by a cluster-aware client and then periodically (every few seconds maybe?) to check that the topology has not changed.

Reply:
Ok. Well, maybe we should stick an enhancement in the backlog somewhere. Every few seconds doesn't seem like a lot, until there are 1000s of clients doing that. Doesn't need to hold up this merge.

Reply:
It isn't really periodic though, If a bucket move occurs or a client otherwise has incorrect slot assignments we need to send MOVED, which will need to calculate the current bucket placement. @jdeppe-pivotal said there is a optimization coming to address this though.